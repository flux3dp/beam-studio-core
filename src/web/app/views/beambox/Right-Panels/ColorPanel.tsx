import React, { useMemo, useState } from 'react';
import { ConfigProvider, InputNumber } from 'antd';

import ColorPicker from 'app/widgets/ColorPicker';
import constant from 'app/actions/beambox/constant';
import HistoryCommandFactory from 'app/svgedit/HistoryCommandFactory';
import storage from 'implementations/storage';
import useDidUpdateEffect from 'helpers/hooks/useDidUpdateEffect';
import { getSVGAsync } from 'helpers/svg-editor-helper';

import styles from './ColorPanel.module.scss';

let svgCanvas;
getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

const deriveElementState = (element: Element) => {
  const fill = element.getAttribute('fill');
  const stroke = element.getAttribute('stroke');
  const strokeWidthAttr = element.getAttribute('stroke-width') || '1';
  const strokeWidth = Number.isNaN(Number(strokeWidthAttr)) ? 1 : Number(strokeWidthAttr);
  return { fill, stroke, strokeWidth };
};

interface Props {
  elem: Element;
}

interface State {
  fill: string;
  stroke: string;
  strokeWidth: number;
}

const ColorPanel = ({ elem }: Props): JSX.Element => {
  const [state, setState] = useState<State>(deriveElementState(elem));
  const { fill, stroke, strokeWidth } = state;
  useDidUpdateEffect(() => {
    setState(deriveElementState(elem));
  }, [elem]);

  const handleFillColorChange = (newColor: string) => {
    const batchCmd = HistoryCommandFactory.createBatchCommand('Color Panel Fill');
    svgCanvas.undoMgr.beginUndoableChange('fill', [elem]);
    svgCanvas.changeSelectedAttributeNoUndo('fill', newColor, [elem]);
    let cmd = svgCanvas.undoMgr.finishUndoableChange();
    if (!cmd.isEmpty()) batchCmd.addSubCommand(cmd);
    svgCanvas.undoMgr.beginUndoableChange('fill-opacity', [elem]);
    svgCanvas.changeSelectedAttributeNoUndo('fill-opacity', newColor === 'none' ? '0' : '1', [
      elem,
    ]);
    cmd = svgCanvas.undoMgr.finishUndoableChange();
    if (!cmd.isEmpty()) batchCmd.addSubCommand(cmd);
    svgCanvas.undoMgr.addCommandToHistory(batchCmd);
    setState({ ...state, fill: newColor });
  };

  const handleStrokeColorChange = (newColor: string) => {
    svgCanvas.changeSelectedAttribute('stroke', newColor, [elem]);
    setState({ ...state, stroke: newColor });
  };

  const handleStrokeWidthChange = (val: number) => {
    svgCanvas.changeSelectedAttribute('stroke-width', val, [elem]);
    setState({ ...state, strokeWidth: val });
  };

  const unit = storage.get('default-units') || 'mm';
  const ratio = useMemo(() => constant.dpmm * (unit === 'inches' ? 25.4 : 1), [unit]);

  return (
    <>
      <div className={styles.block}>
        <div className={styles.label}>Fill</div>
        <ColorPicker allowClear initColor={fill} onChange={handleFillColorChange} />
      </div>
      <div className={styles.block}>
        <div className={styles.label}>Stroke</div>
        <div className={styles.controls}>
          <ColorPicker
            initColor={strokeWidth === 0 ? 'none' : stroke}
            triggerType="stroke"
            onChange={handleStrokeColorChange}
          />
          <ConfigProvider
            theme={{
              components: {
                InputNumber: {
                  controlWidth: 60,
                },
              },
            }}
          >
            <InputNumber
              id="stroke-width"
              value={strokeWidth}
              size="small"
              min={0}
              onChange={handleStrokeWidthChange}
              precision={0}
              // addonAfter={isInch ? 'in' : 'mm'}
              formatter={(value) => `${value / ratio}`}
              parser={(value) => Number(value) * ratio}
            />
          </ConfigProvider>
        </div>
      </div>
    </>
  );
};

export default ColorPanel;
