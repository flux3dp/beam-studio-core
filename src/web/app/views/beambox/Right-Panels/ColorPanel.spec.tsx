import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import ColorPanel from './ColorPanel';

jest.mock('app/actions/beambox/constant', () => ({
  dpmm: 10,
}));

const mockShowColorPicker = jest.fn();
jest.mock('app/actions/dialog-caller', () => ({
  showColorPicker: (...args) => mockShowColorPicker(...args),
}));

const mockCreateBatchCommand = jest.fn();
jest.mock('app/svgedit/HistoryCommandFactory', () => ({
  createBatchCommand: (...args) => mockCreateBatchCommand(...args),
}));
const mockAddSubCommand = jest.fn();
const mockBatchCommand = {
  addSubCommand: (...args) => mockAddSubCommand(...args),
};

jest.mock(
  'app/widgets/Unit-Input-v2',
  () =>
    ({
      id,
      defaultValue,
      getValue,
    }: {
      id: string;
      defaultValue: number;
      getValue: (val: number) => void;
    }) =>
      (
        <div>
          <input
            id={id}
            data-testid={id}
            type="number"
            value={defaultValue}
            onChange={(e) => getValue(parseFloat(e.target.value))}
          />
        </div>
      )
);

const mockGet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => mockGet(...args),
}));

const mockDeleteElements = jest.fn();
jest.mock('app/svgedit/operations/delete', () => ({
  deleteElements: (...args) => mockDeleteElements(...args),
}));

const mockChangeSelectedAttribute = jest.fn();
const mockChangeSelectedAttributeNoUndo = jest.fn();
const mockBeginUndoableChange = jest.fn();
const mockFinishUndoableChange = jest.fn();
const mockAddCommandToHistory = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) => {
    callback({
      Canvas: {
        changeSelectedAttribute: (...args) => mockChangeSelectedAttribute(...args),
        changeSelectedAttributeNoUndo: (...args) => mockChangeSelectedAttributeNoUndo(...args),
        undoMgr: {
          beginUndoableChange: (...args) => mockBeginUndoableChange(...args),
          finishUndoableChange: (...args) => mockFinishUndoableChange(...args),
          addCommandToHistory: (...args) => mockAddCommandToHistory(...args),
        },
      },
    });
  },
}));

const mockElem = {
  getAttribute: jest.fn(),
};

describe('test ColorPanel', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    mockElem.getAttribute
      .mockReturnValueOnce('#ff0000')
      .mockReturnValueOnce('#00ff00')
      .mockReturnValueOnce('1');
    mockGet.mockReturnValueOnce('mm');
    mockCreateBatchCommand.mockReturnValue(mockBatchCommand);
  });

  it('should render correctly', () => {
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(container).toMatchSnapshot();
    expect(mockElem.getAttribute).toHaveBeenCalledTimes(3);
    expect(mockElem.getAttribute).toHaveBeenNthCalledWith(1, 'fill');
    expect(mockElem.getAttribute).toHaveBeenNthCalledWith(2, 'stroke');
    expect(mockElem.getAttribute).toHaveBeenNthCalledWith(3, 'stroke-width');
    expect(mockGet).toHaveBeenCalledTimes(1);
    expect(mockGet).toHaveBeenNthCalledWith(1, 'default-units');
  });

  test('set stroke width should work', () => {
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    const input = container.querySelector('#stroke-width');
    fireEvent.change(input, { target: { value: '2' } });
    expect(mockChangeSelectedAttribute).toHaveBeenCalledTimes(1);
    expect(mockChangeSelectedAttribute).toHaveBeenNthCalledWith(1, 'stroke-width', 20, [mockElem]);
    expect(container).toMatchSnapshot();
  });

  test('set fill color should work', () => {
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(mockShowColorPicker).not.toHaveBeenCalled();
    const fillBlock = container.querySelectorAll('.color')[0];
    fireEvent.click(fillBlock);
    expect(mockShowColorPicker).toHaveBeenCalledTimes(1);
    const opts = mockShowColorPicker.mock.calls[0][0];
    const { originalColor, allowNone, onNewColor } = opts;
    expect(originalColor).toBe('#ff0000');
    expect(allowNone).toBe(true);
    const mockCmd1 = { id: '1', isEmpty: () => false };
    const mockCmd2 = { id: '2', isEmpty: () => false };
    mockFinishUndoableChange.mockReturnValueOnce(mockCmd1).mockReturnValueOnce(mockCmd2);
    onNewColor('#0000ff');
    expect(mockCreateBatchCommand).toBeCalledTimes(1);
    expect(mockCreateBatchCommand).toHaveBeenNthCalledWith(1, 'Color Panel Fill');
    expect(mockBeginUndoableChange).toBeCalledTimes(2);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(1, 'fill', [mockElem]);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(2, 'fill-opacity', [mockElem]);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenCalledTimes(2);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(1, 'fill', '#0000ff', [
      mockElem,
    ]);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(2, 'fill-opacity', '1', [
      mockElem,
    ]);
    expect(mockFinishUndoableChange).toHaveBeenCalledTimes(2);
    expect(mockAddSubCommand).toBeCalledTimes(2);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(1, mockCmd1);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(2, mockCmd2);
    expect(mockAddCommandToHistory).toBeCalledTimes(1);
    expect(mockAddCommandToHistory).toHaveBeenNthCalledWith(1, mockBatchCommand);
  });

  test('set fill color to none should work', () => {
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(mockShowColorPicker).not.toHaveBeenCalled();
    const fillBlock = container.querySelectorAll('.color')[0];
    fireEvent.click(fillBlock);
    expect(mockShowColorPicker).toHaveBeenCalledTimes(1);
    const opts = mockShowColorPicker.mock.calls[0][0];
    const { originalColor, allowNone, onNewColor } = opts;
    expect(originalColor).toBe('#ff0000');
    expect(allowNone).toBe(true);
    const mockCmd1 = { id: '1', isEmpty: () => false };
    const mockCmd2 = { id: '2', isEmpty: () => false };
    mockFinishUndoableChange.mockReturnValueOnce(mockCmd1).mockReturnValueOnce(mockCmd2);
    onNewColor('none');
    expect(mockCreateBatchCommand).toBeCalledTimes(1);
    expect(mockCreateBatchCommand).toHaveBeenNthCalledWith(1, 'Color Panel Fill');
    expect(mockBeginUndoableChange).toBeCalledTimes(2);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(1, 'fill', [mockElem]);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(2, 'fill-opacity', [mockElem]);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenCalledTimes(2);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(1, 'fill', 'none', [
      mockElem,
    ]);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(2, 'fill-opacity', '0', [
      mockElem,
    ]);
    expect(mockFinishUndoableChange).toHaveBeenCalledTimes(2);
    expect(mockAddSubCommand).toBeCalledTimes(2);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(1, mockCmd1);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(2, mockCmd2);
    expect(mockAddCommandToHistory).toBeCalledTimes(1);
    expect(mockAddCommandToHistory).toHaveBeenNthCalledWith(1, mockBatchCommand);
    expect(container).toMatchSnapshot();
  });

  test('set fill to none when stroke is none should delete element', () => {
    mockElem.getAttribute
      .mockReset()
      .mockReturnValueOnce('#ff0000')
      .mockReturnValueOnce('none')
      .mockReturnValueOnce('1');
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(mockShowColorPicker).not.toHaveBeenCalled();
    const fillBlock = container.querySelectorAll('.color')[0];
    fireEvent.click(fillBlock);
    expect(mockShowColorPicker).toHaveBeenCalledTimes(1);
    const opts = mockShowColorPicker.mock.calls[0][0];
    const { originalColor, allowNone, onNewColor } = opts;
    expect(originalColor).toBe('#ff0000');
    expect(allowNone).toBe(true);
    expect(mockDeleteElements).not.toHaveBeenCalled();
    onNewColor('none');
    expect(mockDeleteElements).toHaveBeenCalledTimes(1);
    expect(mockDeleteElements).toHaveBeenNthCalledWith(1, [mockElem]);
  });

  test('set stroke color should work', () => {
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(mockShowColorPicker).not.toHaveBeenCalled();
    const strokeBlock = container.querySelectorAll('.color')[1];
    fireEvent.click(strokeBlock);
    expect(mockShowColorPicker).toHaveBeenCalledTimes(1);
    const opts = mockShowColorPicker.mock.calls[0][0];
    const { originalColor, allowNone, onNewColor } = opts;
    expect(originalColor).toBe('#00ff00');
    expect(allowNone).toBe(true);
    expect(mockChangeSelectedAttribute).not.toHaveBeenCalled();
    onNewColor('#aaaaff');
    expect(mockChangeSelectedAttribute).toHaveBeenCalledTimes(1);
    expect(mockChangeSelectedAttribute).toHaveBeenNthCalledWith(1, 'stroke', '#aaaaff', [mockElem]);
  });

  test('set stroke color to none when fill is none should delete element', () => {
    mockElem.getAttribute
      .mockReset()
      .mockReturnValueOnce('none')
      .mockReturnValueOnce('#00ff00')
      .mockReturnValueOnce('1');
    const { container } = render(<ColorPanel elem={mockElem as any} />);
    expect(mockShowColorPicker).not.toHaveBeenCalled();
    const strokeBlock = container.querySelectorAll('.color')[1];
    fireEvent.click(strokeBlock);
    expect(mockShowColorPicker).toHaveBeenCalledTimes(1);
    const opts = mockShowColorPicker.mock.calls[0][0];
    const { originalColor, allowNone, onNewColor } = opts;
    expect(originalColor).toBe('#00ff00');
    expect(allowNone).toBe(true);
    expect(mockDeleteElements).not.toHaveBeenCalled();
    onNewColor('none');
    expect(mockDeleteElements).toHaveBeenCalledTimes(1);
    expect(mockDeleteElements).toHaveBeenNthCalledWith(1, [mockElem]);
  });
});
