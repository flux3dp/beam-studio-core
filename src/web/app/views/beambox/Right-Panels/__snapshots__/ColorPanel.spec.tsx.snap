// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ColorPanel set fill color to none should work 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      COLOR
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Fill
      </div>
      <div
        class="color none"
        style="background: none;"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke
      </div>
      <div
        class="color"
        style="background: rgb(0, 255, 0);"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke Width
      </div>
      <div>
        <input
          data-testid="stroke-width"
          id="stroke-width"
          type="number"
          value="0.1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`test ColorPanel set stroke width should work 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      COLOR
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Fill
      </div>
      <div
        class="color"
        style="background: rgb(255, 0, 0);"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke
      </div>
      <div
        class="color"
        style="background: rgb(0, 255, 0);"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke Width
      </div>
      <div>
        <input
          data-testid="stroke-width"
          id="stroke-width"
          type="number"
          value="0.1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`test ColorPanel should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      COLOR
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Fill
      </div>
      <div
        class="color"
        style="background: rgb(255, 0, 0);"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke
      </div>
      <div
        class="color"
        style="background: rgb(0, 255, 0);"
      />
    </div>
    <div
      class="block"
    >
      <div
        class="label"
      >
        Stroke Width
      </div>
      <div>
        <input
          data-testid="stroke-width"
          id="stroke-width"
          type="number"
          value="0.1"
        />
      </div>
    </div>
  </div>
</div>
`;
